File: examples/elixir/basic-client/test/runtime_http_test.exs
defmodule RuntimeHttpTest do
  use ExUnit.Case, async: true

  alias RuntimeTypes.HttpRequest

  describe "dispatch/3" do
    test "builds url without query" do
      config = %{base_url: "https://api.example"}
      req = %HttpRequest{
        method: "GET",
        path: "/items",
        query: %{},
        headers: [{"Content-Type", "application/json"}],
        body: ""
      }

      assert {:ok, resp} = RuntimeHttp.dispatch(RuntimeHttpMock, config, req)
      assert resp.status == 200
      assert resp.headers == [{"etag", "\"v1\""}]
      assert resp.body == ~s({"ok":true})
    end

    test "appends query string" do
      config = %{base_url: "https://api.example"}
      req = %HttpRequest{
        method: "GET",
        path: "/items",
        query: %{"verbose" => "true"},
        headers: [],
        body: ""
      }

      assert {:ok, resp} = RuntimeHttp.dispatch(RuntimeHttpMock, config, req)
      assert resp.status == 200
      assert resp.body == ""
    end

    test "propagates client error" do
      config = %{base_url: "https://api.example"}
      req = %HttpRequest{
        method: "GET",
        path: "/fail",
        query: %{},
        headers: [],
        body: ""
      }

      assert {:error, :timeout} == RuntimeHttp.dispatch(RuntimeHttpMock, config, req)
    end
  end

  describe "dispatch exports" do
    test "exports two- and three-arity dispatch" do
      assert {:dispatch, 2} in RuntimeHttp.__info__(:functions)
      assert {:dispatch, 3} in RuntimeHttp.__info__(:functions)
    end
  end
end

File: examples/erlang/basic-client/test/runtime_http_test.erl
-module(runtime_http_test).

-include_lib("eunit/include/eunit.hrl").
-include("runtime_types.hrl").

dispatch_builds_url_without_query_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/items">>,
        query = #{},
        headers = [{<<"Content-Type">>, <<"application/json">>}],
        body = <<>>
    },
    {ok, Resp} = runtime_http:dispatch(runtime_http_mock, Config, Req),
    ?assertEqual(200, Resp#http_response.status),
    ?assertEqual([{<<"etag">>, <<"\"v1\"">>}], Resp#http_response.headers),
    ?assertEqual(<<"{\"ok\":true}">>, Resp#http_response.body).

dispatch_appends_query_string_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/items">>,
        query = #{<<"verbose">> => <<"true">>},
        headers = [],
        body = <<>>
    },
    {ok, Resp} = runtime_http:dispatch(runtime_http_mock, Config, Req),
    ?assertEqual(200, Resp#http_response.status),
    ?assertEqual(<<>>, Resp#http_response.body).

dispatch_propagates_client_error_test() ->
    Config = #{base_url => <<"https://api.example">>},
    Req = #http_request{
        method = <<"GET">>,
        path = <<"/fail">>,
        query = #{},
        headers = [],
        body = <<>>
    },
    ?assertEqual(
        {error, timeout},
        runtime_http:dispatch(runtime_http_mock, Config, Req)
    ).

dispatch_exports_test() ->
    Exports = runtime_http:module_info(exports),
    ?assert(lists:member({dispatch, 2}, Exports)),
    ?assert(lists:member({dispatch, 3}, Exports)).
